### DISCUSSION TITLE
Execute single database query for a GraphQL request

### DISCUSSION THREAD

**Discussion by ankit-joinwal** [Netflix/dgs-framework#502] [Q&A] [answered]
Hello community,
I have a GraphQL service with Neptune GraphDB as data store. Consider SWAPI example for instance:
```
type Film {
  """The title of this film."""
  title: String
   """The director of this film."""
  director: Director

  """The producer(s) of this film."""
  producers: [Producer]
}
```

Suppose I am storing films, directors and producers in a GraphDB such as AWS Neptune as nodes and edges.

I would like to execute single database query for GraphQL Query such as `query films { director { name } producers {name} }`
I do not want to write individual data fetchers for the fields because it would result in multiple database queries. 

Is there a way I can execute single database query for a GraphQL request (based on the fields selected in request) with Netflix DGS ?

**Comment by berngp** [comment, marked as answer]
@ankit-joinwal have you looked at how you can avoid N+1 queries by using [Data Loaders](https://netflix.github.io/dgs/data-loaders/#implementing-a-data-loader).

**Reply by ankit-joinwal** [reply]
Hi @berngp 
Thanks for the quick reply. I am aware of Data loaders but the challenge is that I want to execute a single Gremlin query to make use of Graph traversal otherwise if i use data loaders, there may be cases that the child entity does not exist for the parent entity returned which is undesired. A Graph database like Neptune skips those parent entities for which child entities do not exist . 

What i am trying to build is similar to https://github.com/neo4j-graphql/neo4j-graphql-java/blob/master/examples/dgs-spring-boot/readme.adoc but using gremlin(Neptune) instead of cypher (Neo4j).

The neo4j-graphql-java library executes single Cypher query by translating GraphQL to Cypher.

**Reply by paulbakker** [reply]
I'm not very familiar with Neo/Cypher/Neptune, but this might be helpful: https://netflix.github.io/dgs/advanced/context-passing/#pre-loading

**Reply by ankit-joinwal** [reply]
Thanks @paulbakker Firstly i want to thank the Netflix team for being super active in forums and taking out time to help the community, really appreciate you folks !!!
I think the local context approach can be used to build the solution to the problem I mentioned. In general, i think that pre loading approach makes us write lot many if-else in the parent fetcher but I agree it is one of possible solution.

Thanks again for helping.
